test/bigfloat_dot.jl

```diff
@@ -0,0 +1,107 @@
+# Copyright (c) 2023 MutableArithmetics.jl contributors
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v.2.0. If a copy of the MPL was not distributed with this file, You can obtain
+# one at http://mozilla.org/MPL/2.0/.
+
+backup_bigfloats(v::AbstractVector{BigFloat}) = map(x -> x + 0, v)
+
+absolute_error(accurate::Real, approximate::Real) = abs(accurate - approximate)
+
+function relative_error(accurate::Real, approximate::Real)
+    return absolute_error(accurate, approximate) / abs(accurate)
+end
+
+function dotter(x::V, y::V) where {V<:AbstractVector{<:Real}}
+    let x = x, y = y
+        () -> LinearAlgebra.dot(x, y)
+    end
+end
+
+function reference_dot(x::V, y::V) where {F<:Real,V<:AbstractVector{F}}
+    return setprecision(dotter(x, y), F, 8 * precision(F))
+end
+
+function dot_test_relative_error(x::V, y::V) where {V<:AbstractVector{BigFloat}}
+    buf = MA.buffer_for(LinearAlgebra.dot, V, V)
+
+    input = (x, y)
+    backup = map(backup_bigfloats, input)
+
+    output = BigFloat()
+
+    MA.buffered_operate_to!!(buf, output, LinearAlgebra.dot, input...)
+
+    @test input == backup
+
+    return relative_error(reference_dot(input...), output)
+end
+
+subtracter(s::Real) =
+    let s = s
+        x -> x - s
+    end
+
+our_rand(n::Int, bias::Real) = map(subtracter(bias), rand(BigFloat, n))
+
+function rand_dot_rel_err(size::Int, bias::Real)
+    x = our_rand(size, bias)
+    y = our_rand(size, bias)
+    return dot_test_relative_error(x, y)
+end
+
+function max_rand_dot_rel_err(size::Int, bias::Real, iter_cnt::Int)
+    max_rel_err = zero(BigFloat)
+    for i in 1:iter_cnt
+        rel_err = rand_dot_rel_err(size, bias)
+        <(max_rel_err, rel_err) && (max_rel_err = rel_err)
+    end
+    return max_rel_err
+end
+
+function max_rand_dot_ulps(size::Int, bias::Real, iter_cnt::Int)
+    return max_rand_dot_rel_err(size, bias, iter_cnt) / eps(BigFloat)
+end
+
+function ulper(size::Int, bias::Real, iter_cnt::Int)
+    let s = size, b = bias, c = iter_cnt
+        () -> max_rand_dot_ulps(s, b, c)
+    end
+end
+
+@testset "prec:$prec size:$size bias:$bias" for (prec, size, bias) in
+                                                Iterators.product(
+    (32, 64),
+    (10000, 100000),
```

Comment:
I'm wondering why we need such size for testing

Reply:
I'm not an expert, so the comments I added might make someone who knows better cringe, but I hope it's mostly OK. I based the code and my understanding of the subject matter on the Wikipedia page on Kahan summation and Higham's 1993 paper:

> Higham, Nicholas J. "The accuracy of floating point summation." SIAM Journal on Scientific Computing 14.4 (1993): 783-799.

Reply:
Thanks for adding the comments, could you add the ref to Higham's paper as well then ? The algorithm is a bit cryptic without a reference :grin: